lib/lintSource.js

```diff
@@ -34,74 +34,60 @@ module.exports = function lintSource(stylelint, options = {}) {
 		if (isCodeNotFile && isPathNotFoundError(err)) return false;
 
 		throw err;
 	});
 
-	return getIsIgnored.then((isIgnored) => {
-		if (isIgnored) {
-			/** @type {PostcssResult} */
-
-			return options.existingPostcssResult
-				? Object.assign(options.existingPostcssResult, {
-						stylelint: createEmptyStylelintPostcssResult(),
-				  })
-				: createEmptyPostcssResult(inputFilePath);
-		}
+	if (isIgnored) {
+		return options.existingPostcssResult
+			? Object.assign(options.existingPostcssResult, {
+					stylelint: createEmptyStylelintPostcssResult(),
+			  })
+			: createEmptyPostcssResult(inputFilePath);
+	}
 
-		const configSearchPath = stylelint._options.configFile || inputFilePath;
+	const configSearchPath = stylelint._options.configFile || inputFilePath;
 
-		const getConfig = stylelint.getConfigForFile(configSearchPath, inputFilePath).catch((err) => {
+	const configForFile = await stylelint
+		.getConfigForFile(configSearchPath, inputFilePath)
+		.catch((err) => {
 			if (isCodeNotFile && isPathNotFoundError(err))
 				return stylelint.getConfigForFile(process.cwd());
 
 			throw err;
 		});
 
-		return getConfig.then((result) => {
-			if (!result) {
-				throw new Error('Config file not found');
-			}
-
-			const config = result.config;
-			const existingPostcssResult = options.existingPostcssResult;
-			const stylelintResult = {
-				ruleSeverities: {},
-				customMessages: {},
-				disabledRanges: {},
-			};
-
-			if (existingPostcssResult) {
-				const stylelintPostcssResult = Object.assign(existingPostcssResult, {
-					stylelint: stylelintResult,
-				});
-
-				return lintPostcssResult(stylelint._options, stylelintPostcssResult, config).then(
-					() => stylelintPostcssResult,
-				);
-			}
-
-			return stylelint
-				._getPostcssResult({
-					code: options.code,
-					codeFilename: options.codeFilename,
-					filePath: inputFilePath,
-					codeProcessors: config.codeProcessors,
-					customSyntax: config.customSyntax,
-				})
-				.then((postcssResult) => {
-					const stylelintPostcssResult = Object.assign(postcssResult, {
-						stylelint: stylelintResult,
-					});
-
-					return lintPostcssResult(stylelint._options, stylelintPostcssResult, config).then(
-						() => stylelintPostcssResult,
-					);
-				});
-		});
+	if (!configForFile) {
+		return Promise.reject(new Error('Config file not found'));
+	}
+
+	const config = configForFile.config;
+	const existingPostcssResult = options.existingPostcssResult;
+	const stylelintResult = {
+		ruleSeverities: {},
+		customMessages: {},
+		disabledRanges: {},
+	};
+
+	const postcssResult =
+		existingPostcssResult ||
+		(await stylelint._getPostcssResult({
+			code: options.code,
+			codeFilename: options.codeFilename,
+			filePath: inputFilePath,
+			codeProcessors: config.codeProcessors,
+			customSyntax: config.customSyntax,
+		}));
+
+	const stylelintPostcssResult = Object.assign(postcssResult, {
+		stylelint: stylelintResult,
```

Comment:
This whole section is much simpler now. 👍 

Style nitpick, feel free to ignore. Could change to something like:

```js
const stylelintPostcssResult = {
  ...postcssResult,
  stylelint: {
    ruleSeverities: {},
    customMessages: {},
    disabledRanges: {},
  },
};
```
And then remove that `stylelintResult` from a few lines above?

Reply:
I'm worried that code could rely on object mutation. I've seen it in some places. `Object.assign` will change original object, while spread will create a new one.

Reply:
👍 makes sense.